### DISCUSSION TITLE
Why cmark-gfm instead of jgm / commonmark-hs?

### DISCUSSION THREAD

**Discussion by digitalmoksha** [kivikakk/comrak#254] [General]
I'm curious.  What about porting from https://github.com/jgm/commonmark-hs rather than cmark-gfm? jgm as [speced out extensions](https://github.com/jgm/commonmark-hs/tree/master/commonmark-extensions/test) pretty well, it seems very extensible and supports source positions for all elements. And supports GFM.

**Comment by gjtorikian** [comment]
I'm not sure what @kivikakk's reply will be, but I can guess: we were both at GitHub when the GFM spec came out; and indeed, @kivikakk worked on that very spec.

This lib has gone on to evolve indepentently of GitHub/cmark-gfm and I'm sure any extension / suggestion that's PRed would be merged, a la https://github.com/kivikakk/comrak/issues/251

**Reply by kivikakk** [reply]
Right, yeah! Comrak's hit full base CommonMark spec compliance 11 months before commonmark-hs's initial commit. At the time I was primary maintainer of GitHub's cmark-gfm so it was easy for me to make updates in lockstep. And as @gjtorikian says, I basically don't turn down thoughtful contributions as long as they don't impact default compatibility.

**Reply by digitalmoksha** [reply]
That makes total sense. And I must say awesome job you guys did on `cmark-gfm` and `commonmarker`.

Is `comrak` now fully compatible with `cmark-gfm`, including `data-sourcepos`? Or is there something left to do?

I don't know Rust (yet), but I was looking through the code. I didn't see a direct extension mechanism - for example description list support seems directly inline in the code. Is that how you would see adding say math support?  `cmark-gfm` has their `extensions` directory, though I admit I never fully understood their mechanism.

It also looks like tests are hardcoded, as opposed to using example files that CommonMark and GFM uses, though I may have overlooked that.

Sorry for the pointed questions - I've been using `cmark-gfm` and `commonmarker` for awhile, but starting to cast my eye around to other possibilities.

**Reply by kivikakk** [reply]
`sourcepos` is semi-complete; I have some unfinished WIP at #232. Spec-wise Comrak is fully compatible, but some lesser-used features of the software itself are missing. (XML output is a good example, also WIP in #232.) On the other hand, it has many more contributed features.

Extensions aren't loaded at runtime or anything like that, no, and I would add new features (such as math) the same way. `cmark-gfm` is _mostly_ the same: extensions are built-in at compile time, but the design of the code there allows extensions to hook themselves in and thus be separated out into a different source directory/library object archive before linking.

Comrak has its unit tests, which allow for fast iteration and Rust API tests, but it's also tested in CI against all the same example files used by `cmark-gfm`; see [`script/cibuild`](https://github.com/kivikakk/comrak/blob/main/script/cibuild) and [a typical CI run](https://github.com/kivikakk/comrak/actions/runs/3634894701).

**Reply by kivikakk** [reply]
Also, no need to apologise for the questions, hah :) Happy to answer them! I don't really have much time to dedicate to Comrak these days, unfortunately — it's been many years since I was actively working in the CommonMark space, or with Rust for that matter, and life is rather a lot to manage these days — but I continue to do my best to keep it maintained and secure, encourage/shepherd new contributors, and make small improvements over time myself.

**Reply by digitalmoksha** [reply]
> life is rather a lot to manage these days

Oh yeah, no kidding. Totally understand.

> it's also tested in CI against all the same example files used by cmark-gfm

Ah, I knew I had to be missing a piece. 👍 

> `sourcepos` is semi-complete

This is actually a requirement for us - it's used both in our backend and frontend. With my limited Rust knowledge, I'm not sure I'd be much help right now.

> Extensions aren't loaded at runtime or anything like that, no, and I would add new features (such as math) the same way. cmark-gfm is mostly the same: extensions are built-in at compile time, but the design of the code there allows extensions to hook themselves in and thus be separated out into a different source directory/library object archive before linking.

I have to admit I like the ability to have extensions hook themselves in, and be separated out.  Looking at [rlidwka/markdown-it.rs](https://github.com/rlidwka/markdown-it.rs/tree/master/src/plugins), that's what they are doing.  Of course `markdown-it` is specifically built that way originally.

I'm thinking in terms of being able to create our own crate that pulls in yours and then adds an extension, such as our [multiline block quote](https://docs.gitlab.com/ee/user/markdown.html#multiline-blockquote), or our [task list extension](https://docs.gitlab.com/ee/user/markdown.html#task-lists).  Things that might not necessarily belong in this repo (though I would love it if it did, enabled via a `glfm` flag 😄). We currently handle those things before/after the markdown transformation.

**Reply by kivikakk** [reply]
> This is actually a requirement for us - it's used both in our backend and frontend. With my limited Rust knowledge, I'm not sure I'd be much help right now.

I understand! I have a personal todo to finish it off after I've finished moving house, but that's unlikely to be for a month or two.

> I'm thinking in terms of being able to create our own crate that pulls in yours and then adds an extension, such as our [multiline block quote](https://docs.gitlab.com/ee/user/markdown.html#multiline-blockquote), or our [task list extension](https://docs.gitlab.com/ee/user/markdown.html#task-lists).

I'd be very happy for such things to be possible! I think it shouldn't be particularly complicated to make happen; just needs some care to make sure the extension API that results is useful in the right ways. (Some more of my thoughts about how to do this are at  https://github.com/kivikakk/comrak/issues/258#issuecomment-1356182880.)

**Reply by kivikakk** [reply]
For what it's worth, sourcepos landed in [0.18.0](https://github.com/kivikakk/comrak/releases/0.18.0).

**Reply by digitalmoksha** [reply]
Oh, very awesome!! Thanks 🚀
